presto-main/src/main/java/com/facebook/presto/util/Failures.java

```diff
@@ -71,28 +92,36 @@ public static ExecutionFailureInfo toFailure(Throwable failure)
             remoteHost = ((PrestoTransportException) failure).getRemoteHost();
         }
 
+        Throwable cause = failure.getCause();
+        if (cause != null && seenFailures.contains(cause)) {
+            log.warn(cause, "found cyclic cause");
+            cause = null;
+        }
+        else {
+            seenFailures.add(cause);
+        }
         return new ExecutionFailureInfo(type,
                 failure.getMessage(),
-                toFailure(failure.getCause()),
-                toFailures(asList(failure.getSuppressed())),
+                toFailure(cause, seenFailures),
+                toFailures(asList(failure.getSuppressed()), seenFailures),
                 Lists.transform(asList(failure.getStackTrace()), toStringFunction()),
                 getErrorLocation(failure),
                 toErrorCode(failure),
                 remoteHost);
     }
 
-    public static void checkCondition(boolean condition, ErrorCodeSupplier errorCode, String formatString, Object... args)
+    private static List<ExecutionFailureInfo> toFailures(Collection<? extends Throwable> failures, Set<Throwable> seenFailures)
     {
-        if (!condition) {
-            throw new PrestoException(errorCode, format(formatString, args));
+        ImmutableList.Builder<ExecutionFailureInfo> failureInfoBuilder = ImmutableList.builder();
+        for (Throwable failure : failures) {
+            if (seenFailures.contains(failure)) {
+                log.warn(failure, "found cyclic suppression");
```

Comment:
technically, this can happen also for non-cyclic structure, when `toFailures(Collection<? extends Throwable> failures)` is called (repeated element, or two elements sharing suppression) or when `toFailure(Throwable failure)` is called (diamond suppression graph)

Also, is it safe to ask `log.warn` to log the `failure`, it it contains cyclic suppression? Will the logger handle it?

Reply:
- You are right it doesn't have to be a cycle. To implement a cycle detection for this may be an overkill though.
- Interestingly, yes. Check `java.lang.Throwable.printStackTrace`

Reply:
I don't think we should log it. I find no value in logging.